src/Query/DateRange.ts

```diff
@@ -0,0 +1,63 @@
+import moment from 'moment';
+
+export class DateRange {
+    start: moment.Moment;
+    end: moment.Moment;
+
+    constructor(start: moment.Moment, end: moment.Moment) {
+        this.start = start;
+        this.end = end;
+
+        if (end.isBefore(start)) {
+            this.start = end;
+            this.end = start;
+        }
+
+        // Dates shall be at midnight eg 00:00
+        this.start = this.start.startOf('day');
+        this.end = this.end.startOf('day');
+    }
+
+    public static buildRelative(range: moment.unitOfTime.StartOf) {
+        // Treat all weeks as ISO 8601 weeks
+        const unitOfTime = range === 'week' ? 'isoWeek' : range;
+
+        return new DateRange(moment().startOf(unitOfTime).startOf('day'), moment().endOf(unitOfTime).startOf('day'));
+    }
+
+    public static buildSpecific(date: string, dateFormat: string, range: moment.unitOfTime.StartOf | null): DateRange {
+        return new DateRange(moment(date, dateFormat).startOf(range), moment(date, dateFormat).endOf(range));
+    }
+
+    public static buildInvalid(): DateRange {
+        return new DateRange(moment.invalid(), moment.invalid());
+    }
+
+    public isValid(): boolean {
+        return this.start.isValid() && this.end.isValid();
```

Comment:
From reading the code, I wonder if line 37 means that this PR is not a pure refactoring?

I'm asking because the original code in DateField seems to be using OR not AND:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/d6747d7c1555bb42e753ed7da7266502a864660b/src/Query/Filter/DateField.ts#L57-L58

Reply:
Actually `!A || !B` is logically equivalent to `!(A && B)` (Check in the new code is done on `!fieldDates.isValid()`)

I will explain in detail when we pair

Reply:
Oops - fair point!!